onnxruntime/core/providers/cpu/tensor/upsample.cc

```diff
@@ -408,39 +410,41 @@ void UpsampleBilinear(int64_t batch_size,
       dx2[x] = 0.5f;
     }
   }
-
-  for (int64_t n = 0; n < batch_size; ++n) {
-    for (int64_t c = 0; c < num_channels; ++c) {
-      for (int64_t y = 0; y < output_height; ++y) {
-        for (int64_t x = 0; x < output_width; ++x) {
-          // when use_extrapolation is set and original index of x or y is out of the dim range
-          // then use extrapolation_value as the output value.
-          if (use_extrapolation &&
-              ((y_original[y] < 0 || y_original[y] > static_cast<float>(input_height - 1)) ||
-               (x_original[x] < 0 || x_original[x] > static_cast<float>(input_width - 1)))) {
-            Ydata[output_width * y + x] = static_cast<T>(extrapolation_value);
-            continue;
-          }
-
-          // subscript ordering in the variable - (xy)
-          T X11 = Xdata[input_width_mul_y1[y] + in_x1[x]];
-          T X21 = Xdata[input_width_mul_y1[y] + in_x2[x]];
-          T X12 = Xdata[input_width_mul_y2[y] + in_x1[x]];
-          T X22 = Xdata[input_width_mul_y2[y] + in_x2[x]];
-
-          Ydata[output_width * y + x] = static_cast<T>(dx2[x] * dy2[y] * X11 +
-                                                       dx1[x] * dy2[y] * X21 +
-                                                       dx2[x] * dy1[y] * X12 +
-                                                       dx1[x] * dy1[y] * X22);
-        }
-      }
-      Xdata += input_height * input_width;
-      Ydata += output_width * output_height;
-    }
+  
+ for (int64_t n = 0; n < batch_size; ++n) {
+    concurrency::ThreadPool::TrySimpleParallelFor(tp, num_channels,
+                                                  [&](std::ptrdiff_t c) {
```

Comment:
Only technical comment on the PR is whether we need to back off to a sequential execution when the output images are tiny?  Given that we are upsampling, is it OK to assume that we will not encounter that case?  One way to get sequential execution without having two branches of the code is to pass in tp=nullptr.  [ Ultimately we'd like this to be handled in the thread pool cost model, but per e-mail discussion, it looks like we need some investigation and tuning there .]

Reply:
added a condition that it only parallelize if output size is greater than 64.